packages/contractkit/src/identity/offchain/schema-utils.ts

```diff
@@ -21,62 +26,228 @@ export class OffchainError extends RootError<SchemaErrorTypes.OffchainError> {
+
+    return this.deserialize(rawData.result)
+  }
+
+  async readEncrypted(
+    senderAddress: string,
+    recipientAddress: Address
+  ): Promise<Result<DataType, SchemaErrors>> {
+    const result = await readEncrypted(this.wrapper, this.dataPath, senderAddress, recipientAddress)
+    if (!result.ok) {
+      return Err(result.error)
+    }
+
+    return this.deserialize(result.result)
   }
 }
 
-export const readWithSchemaAsResult = async <T>(
-  wrapper: OffchainDataWrapper,
-  type: t.Type<T>,
-  account: Address,
-  dataPath: string
-): Promise<Result<T, SchemaErrors>> => {
-  const resp = await wrapper.readDataFromAsResult(account, dataPath)
+export class BinarySchema {
+  constructor(readonly wrapper: OffchainDataWrapper, readonly dataPath: string) {}
 
-  if (!resp.ok) {
-    return Err(new OffchainError(resp.error))
+  async write(data: Buffer) {
+    await this.wrapper.writeDataTo(data, this.dataPath)
   }
 
-  try {
-    const asJson = JSON.parse(resp.result)
-    const parseResult = type.decode(asJson)
-    if (isRight(parseResult)) {
-      return Ok(parseResult.right)
+  async writeEncrypted(data: Buffer, fromAddress: string, toAddress: string) {
+    return writeEncrypted(this.wrapper, this.dataPath, data, fromAddress, toAddress)
+  }
+
+  async writeWithSymmetric(data: Buffer, fromAddress: string, toAddresses: string[]) {
+    return writeEncryptedWithSymmetric(this.wrapper, this.dataPath, data, fromAddress, toAddresses)
+  }
+
+  async read(account: string): Promise<Result<Buffer, SchemaErrors>> {
+    const rawData = await this.wrapper.readDataFromAsResult(account, this.dataPath)
+    if (!rawData.ok) {
+      return Err(new OffchainError(rawData.error))
     }
-    return Err(new RootError(SchemaErrorTypes.InvalidDataError))
-  } catch (error) {
-    return Err(new RootError(SchemaErrorTypes.InvalidDataError))
+
+    return Ok(rawData.result)
+  }
+
+  async readEncrypted(senderAddress: string, recipientAddress: Address) {
+    return readEncrypted(this.wrapper, this.dataPath, senderAddress, recipientAddress)
   }
 }
 
-export const readWithSchema = makeAsyncThrowable(readWithSchemaAsResult)
+// label = PRF(ECDH(A, B), A || B || data path)
+// ciphertext path = "/cosmetic path/" || base64(label)
+export function getDataPath(
+  path: string,
+  sharedSecret: Buffer,
+  senderPublicKey: string,
+  receiverPublicKey: string
+) {
+  const senderPublicKeyBuffer = Buffer.from(trimLeading0x(senderPublicKey), 'hex')
+  const receiverPublicKeyBuffer = Buffer.from(trimLeading0x(receiverPublicKey), 'hex')
 
-export const writeWithSchema = async <T>(
+  return createHmac('sha3-256', sharedSecret)
+    .update(Buffer.concat([senderPublicKeyBuffer, receiverPublicKeyBuffer, Buffer.from(path)]))
+    .digest('base64')
+}
+
+export const writeEncrypted = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  data: Buffer,
+  fromAddress: string,
+  toAddress: string
+) => {
+  const [fromPubKey, toPubKey] = await Promise.all([
```

Comment:
Unless I'm mistaken this is pretty critical:

I.e. we should not be using the account addresses to compute the corresponding public keys and thus shared secrets. The reason being that any of the accounts might represent smart contracts and thus are not able to use the corresponding private key. Instead we should always be retrieving an accounts `dataEncryptionKey` with a future improvement being that such key can be registered off-chain